hoot-core/src/main/cpp/hoot/core/io/OsmGbdxXmlWriter.cpp

```diff
@@ -0,0 +1,704 @@
+#include <QDateTime>
+#include <QFile>
+#include <QXmlStreamWriter>
+
+using namespace geos::geom;
+using namespace std;
+
+namespace hoot
+{
+
+unsigned int OsmGbdxXmlWriter::logWarnCount = 0;
+
+HOOT_FACTORY_REGISTER(OsmMapWriter, OsmGbdxXmlWriter)
+
+OsmGbdxXmlWriter::OsmGbdxXmlWriter() :
+_formatXml(ConfigOptions().getOsmMapWriterFormatXml()),
+_precision(ConfigOptions().getWriterPrecision()),
+_encodingErrorCount(0),
+_fileNumber(0)
+{
+}
+
+OsmGbdxXmlWriter::~OsmGbdxXmlWriter()
+{
+  close();
+}
+
+QString OsmGbdxXmlWriter::removeInvalidCharacters(const QString& s)
+{
+  // See #3553 for an explanation.
+
+  QString result;
+  result.reserve(s.size());
+
+  bool foundError = false;
+  for (int i = 0; i < s.size(); i++)
+  {
+    QChar c = s[i];
+    // See http://stackoverflow.com/questions/730133/invalid-characters-in-xml
+    if (c < 0x20 && c != 0x9 && c != 0xA && c != 0xD)
+    {
+      foundError = true;
+    }
+    else
+    {
+      result.append(c);
+    }
+  }
+
+  if (foundError)
+  {
+    _encodingErrorCount++;
+    if (logWarnCount < Log::getWarnMessageLimit())
+    {
+      LOG_WARN("Found an invalid character in string: '" << s << "'");
+      LOG_WARN("  UCS-4 version of the string: " << s.toUcs4());
+    }
+    else if (logWarnCount == Log::getWarnMessageLimit())
+    {
+      LOG_WARN(className() << ": " << Log::LOG_WARN_LIMIT_REACHED_MESSAGE);
+    }
+    logWarnCount++;
+  }
+
+  return result;
+}
+
+void OsmGbdxXmlWriter::open(QString url)
+{
+//  if (url.toLower().endsWith(".gxml"))
+//  {
+//    url.remove(url.size() - 5, url.size());
+//  }
+
+//  _outputDir = QDir(url);
+//  _outputDir.makeAbsolute();
+//  _outputFileName = _outputDir.dirName();
+
+  QFileInfo fi(url);
+  _outputDir = fi.absoluteDir();
+  _outputFileName = fi.baseName();
+
+  if (_outputDir.exists() == false)
+  {
+    if (QDir().mkpath(_outputDir.path()) == false)
+    {
+      throw HootException("Error creating directory for writing.");
+    }
+  }
+  _bounds.init();
+}
+
+void OsmGbdxXmlWriter::_newOutputFile()
+{
+  // Close the old file and open a new one
+  if (_fp.get())
+  {
+    // Calling this so that the XML elements & document get closed
+    close();
+  }
+
+  // The output has had a few changes.....
+//  QString url = _outputDir.filePath(UuidHelper::createUuid().toString().replace("{", "").replace("}", "") + ".xml");
+//  QString url = _outputDir.filePath(QString("%1_%2.xml").arg(_outputFileName,QString::number(_fileNumber++).rightJustified(4,'0')));
+//  QString url = _outputDir.filePath(QString("%1_%2.xml").arg(_outputFileName).arg(_fileNumber++));
+  QString url = _outputDir.filePath(QString("%1_00_%2.xml").arg(_outputFileName).arg(_fileNumber++));
+
+  // If the file exists, increment the middle _00_ in the name.
+  // NOTE: This assumes that there can be a maximum of 10 copies of a filename....
+  if (QFile::exists(url))
+  {
+//    LOG_ERROR("Clash: Orig Filename: " + url);
+    int inc = 0;
+    while (QFile::exists(url)) {
```

Comment:
c++ curmudgeon says this curly brace needs to be on its own line!

Reply:
Fixed